File: arowolo/arrays/day-7/subsetsII.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
 var subsetsWithDup = function(nums) {
    let prevIndex = 0;
   nums.sort((a,b)=>a-b);
   
   let res = [[]];
   for (let i=0;i<nums.length;i++) {    
       let l = res.length;
       for (let j=prevIndex;j<l;j++) res.push([...res[j], nums[i]]);
       if (nums[i+1] == nums[i]) prevIndex = l;
       else prevIndex = 0;
   }
   
   return res;
   
};

File: Tope/Day7/subsetsII.py
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = 1 << len(nums)
        result = []
        for i in range(0, n):
            subset = self.convert(i, nums)
            result.append(tuple(sorted(subset)))

        result = set(result)
        return [list(entries) for entries in result]

    def convert(self, i, nums):
        k = i
        index = 0
        res = []
        while k > 0:
            if (k & 1) == 1:
                res.append(nums[index])

            k >>= 1
            index += 1
        return res